### PR TITLE
added: wireless default unit

### PR DESCRIPTION
**Pull request by jeff-cai**
#4

Preliminary wireless unit.
Future development on displaying ip to be considered.

### CODE REVIEW

**Comment by qdbp** [comment]
I would encourage the direct use of /sys/ and /proc/ over requiring binaries, both for performance and compatibility. 

One of the (hitherto unstated) stretch goals for this project is to have no binary blob dependencies (with the exception of units like PY3NVGPU which monitor a proprietary subsystem).

**Comment by jeff-cai** [comment]
Further development will migrate towards reading from /sys/ and /proc/. In it's current form I think of it as a placeholder unit that returns the intended outcome (mostly for my own uses) that I thought was stable enough to merge to master despite requiring dependencies.

**Comment by jeff-cai** [comment]
The [arch wiki](https://wiki.archlinux.org/index.php/Wireless_network_configuration#Manual_setup) also seems to suggest that these packages are the standard for managing wireless networks, so from a compatibility side I think it would be safe. Though it does mention:

>  wireless_tools - _currently deprecated, but still widely supported._

Perhaps updating to depend on `iw` would be better future-proofing.

File: py3s_default_units.py [hunk 1]
```diff
@@ -197,6 +197,54 @@ def get_chunk(self):
...
+    '''
+    wireless network information
+    SSID, quality
+
+    Requires:
+        iw (probably?)
+        wireless-tools?
+    '''
+
+    def __init__(self, *args, wlan_id='wlp2s0', **kwargs):
```
Comment (qdbp) [inline py3s_default_units.py]: Do not guess wlan id. It should be a positional argument.

File: py3s_default_units.py [hunk 2]
```diff
@@ -197,6 +197,54 @@ def get_chunk(self):
 
         return output
 
+class PY3Wireless(PY3Unit):
+    '''
+    wireless network information
+    SSID, quality
+
+    Requires:
+        iw (probably?)
```
Comment (qdbp) [inline py3s_default_units.py]: Ascertain which ones.

**Review by qdbp — CHANGES_REQUESTED** [review]

**Comment by qdbp** [comment]
Please separate the `get_chunk` method into `read` and `format`, as per the new API.

File: src/py3s_default_units.py [hunk 3]
```diff
@@ -300,8 +300,11 @@ def format(self, output):
 class PY3Wireless(PY3Unit):
     """Provide wireless network information.
 
-    Outputs:
-        SSID, quality
+    Output API:
+        'status': up, down, none
```
Comment (qdbp) [inline src/py3s_default_units.py]: please add type prefixes, i.e. 'SSID -> s_SSID', etc

File: src/py3s_default_units.py [hunk 4]
```diff
@@ -311,42 +314,51 @@ def __init__(self, wlan_id, *args, **kwargs):
...
+            output['status'] = "down" 
+            return output
         # Not connected case
         if 'off/any' in out: # if not connected: "ESSID:off/any"
-            return "w: none"
+            output['status'] = "none"
+            return output
 
         # Raw output data
         raw_SSID = findall('ESSID:"(.*?)"', out)[0]
```
Comment (qdbp) [inline src/py3s_default_units.py]: does this add a leading space to the returned value of SSID?

File: src/py3s_default_units.py [hunk 5]
```diff
@@ -311,42 +314,51 @@ def __init__(self, wlan_id, *args, **kwargs):
...
         # SSID
-        SSID = raw_SSID
-        # Maybe do something with SSID later,
-        # Perhaps display 'home' if connected to home SSID
+        output['SSID'] = raw_SSID
 
         # Quality (overall quality of the link)
         raw_quality = int(raw_quality_num) / int(raw_quality_denom)*100
-        quality = "{:3.0f}%".format(raw_quality)
+        output['quality'] = "{:3.0f}%".format(raw_quality)
```
Comment (qdbp) [inline src/py3s_default_units.py]: should output raw floats, not formatted strings; formatting is `format`'s job!

**Review by qdbp — CHANGES_REQUESTED** [review]
highlighted some problems

File: src/py3s_default_units.py [hunk 6]
```diff
@@ -296,6 +298,73 @@ def format(self, output):
                 .format(braces[0], pct_str, braces[1], rem_string, status_string))
 
 
+class PY3Wireless(PY3Unit):
+    """Provide wireless network information.
+
+    Output API:
+        's_status': up, down, none
+        's_SSID'
```
Comment (qdbp) [inline src/py3s_default_units.py]: be more descriptive: SSID of what? the currently connected-to network? one with the strongest signal? etc.

File: src/py3s_default_units.py [hunk 7]
```diff
@@ -296,6 +298,73 @@ def format(self, output):
...
+
+    def __init__(self, wlan_id, *args, **kwargs):
+
+        """
+        Args:
+            wlan_id:
+                wireless device name from `ip link`
+                example: wlan0
+        """
+        self.wlan_id=wlan_id
```
Comment (qdbp) [inline src/py3s_default_units.py]: PEP8. (applies in other places too)

File: src/py3s_default_units.py [hunk 8]
```diff
@@ -296,6 +298,73 @@ def format(self, output):
...
+        super().__init__(*args, requires=['iwconfig'], **kwargs)
+
+    def read(self):
+        # Future: read stats from /proc/net/wireless?
+        # Raw
+        out = check_output(['iwconfig', self.wlan_id]).decode('ascii')
+        line1 = out.split('\n')[0]
+
+        output = {'s_status': "up", "s_SSID": "SSID", "f_quality": 0}
```
Comment (qdbp) [inline src/py3s_default_units.py]: "SSID" is a silly default. I recommend following the convention of the other unit and introducing some `b_` error flags for when this unit's output is not applicable; i.e. there is no connection.

**Review by qdbp — CHANGES_REQUESTED** [review]

**Pull request merged**